php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+payload using [method reflection](https://www.php.net/manual/en/class.reflectionfunctionabstract.php) mechanism provided by php.    
+
+The invocation context and heartbeats could be addressed `ContextInterface` object. The error handling done using classic
+PHP exceptions.
+
+```php
+function processFile(string $url): string
+{
+    $result = doSomeWork($url);
+
+    return $result;
+}
+```
+
+Heartbeat example:
+
+```php
+use Temporal\Workflow\ContextInterface;
+use Temporal\Workflow\Activity;
+
+// parameter order does not matter
+function downloadFile(ContextInterface $ctx, string $url)
+{
+    $downloader = new FileDownloader();
+    $downloader->onProgress(function($progress) use($ctx){
+        // subject of discussion
+        Activity::heartbeat($ctx, $progress);
+    });
+
+    return $downloader->download($url);
+}
+```
+
+Alternatively we can embed the heartbeat method directly into `ContextInterface` or `ActivityContextInterface`:
+
+```php
+use Temporal\Workflow\ActivityContextInterface;
+
+// parameter order does not matter
+function downloadFile(ActivityContextInterface $ctx, string $url)
+{
+    $downloader = new FileDownloader();
+    $downloader->onProgress(function($progress) use($ctx){
+        // subject of discussion
+        $ctx->heartbeat($progress);
+    });
+
+    // for external activities
+    dump($ctx->getTaskToken());
+
+    return $downloader->download($url);
+}
+```
+
+The second approach deviate from a read-only concept of contexts but makes testing much easier.
+
+### Registration
+Activities can be registered in corresponding workers. Since workflows will run in different process there are no need to
+mix two abstractions together:
+
+```php
+$worker = new Activities();
+$worker->register('activity.name', 'functionName');
+
+// for objects
+$worker->register('activity.name2', [$object, 'methodName']);
+
+// blocking
+$worker->run(new RoadRunner\Worker(...));
+```
+
+> It might be beneficial to provide underlying registry abstraction to simplify the coding of workflows. See the
+> example below.
+
+### Using Annotations
+In a more sophisticated frameworks, some activities can be detected and registered automatically using static analysis
+of codebase and annotations. No common interface required.
+
+```php
+class FileController 
+{
+    /**
+     * @Workflow\Activity("name" = "file.download") 
+     */         
+    public function processFile(string $file): string 
+    {
+        $result = $this->doProcessing($file);
+
+        return $result;
+    }
+
+    // ...
+}
+```
+
+> PHP 8 annotations or alternative registration methods possible as well.
+
+### Logging
+It is possible to provide access to log stream from the `ActivityContextInterface` which makes possible to use PSR-3 compatible
+loggers.
+
+```php
+$logger = new ActivityLogger($ctx);
+$logger->debug('Hello!');
+
+// or directly
+$ctx->log('debug', 'message', ['context']);
+```
+
+### Payloads
```

Comment:
Is it possible to serialize them to JSON seamlessly? Or make serialization pluggable as it is done in Go and Java?

Reply:
Absolutely, it will be controlled on the coordinator end so we can change serialization mechanisms at any time. Payload reqs are the same as in other languages - be serializable, do not carry resources and etc.